<a href="https://colab.research.google.com/github/zzaeyun/ESAA22-2/blob/main/%ED%8C%8C%EC%9D%B4%EC%8D%AC%EC%99%84%EB%B2%BD%EA%B0%80%EC%9D%B4%EB%93%9C_39to86.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **04 데이터 핸들링 - 판다스**
 

---
- 2차원 데이터는 효과적으로 데이터를 담을 수 있음
- 판다스는 2차원 데이터를 효율적으로 가공/처리할 수 있는 기능 제공(넘파이보다 유연)
- 판다스는 CSV등의 파일을 쉽게 DataFrame으로 변경 가능

**DataFrame**
- 2차원 데이터를 담는 구조체
- Index: 개별 데이터를 고유하게 식별하는 Key
- Series: 칼럼이 하나인 데이터 구조체


**판다스 시작 - 파일을 DataFrame으로 로딩, 기본 API**
- import pandas as pd
- 캐글에서 제공하는 데이터를 사용, csv파일
- read_csv(), CSV: 칼럼을 ','(콤마)로 구분
- read_table(): 칼럼을 '\t'(탭)으로 구분

**read_csv**
- filepath: 로드하려는 데이터 파일의 경로를 포함한 파일명입력
- read_csv('파일명', sep='\t')=read_table()
- 파일명 인자로 들어온 파일을 로딩해 DataFrame 객체로 봔환

**DataFrame**
- 데이터 파일의 첫 번째 줄에 있던 칼럼 문자열이 DataFrame의 칼럼으로 할당 (파라미터 지정이 없으면 맨 처음 로우를 칼럼명으로 인지)
- Index 값을 0, 1, 2, ... 지정, 고유의 레코드를 식별하는 역할


In [42]:
import pandas as pd

In [43]:
titanic_df=pd.read_csv('titanic_train.csv')

In [44]:
titanic_df.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S

In [45]:
print('titanic 변수 type:', type(titanic_df))
titanic_df

titanic 변수 type: <class 'pandas.core.frame.DataFrame'>


PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0              1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1              2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2              3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S
3              4         1       1  Futrelle, M...  female  35.0      1      0          113803  53.1000  C123        S
4              5         0       3  Allen, Mr. ...    male  35.0      0      0          373450   8.0500   NaN        S
..           ...       ...     ...             ...     ...   ...    ...    ...             ...      ...   ...      ...
886          887         0       2  Montvila, R...    male  27.0      0      0          211536  13.0000   NaN        S
887          888         1       1  Graham, Mis...  female  19.0      0      0          112053  30.0000   B42        S
888          889         0       3  Johnston, M...  female   NaN      1      2      W./C. 6607  23.4500   NaN        S
889          890         1       1  Behr, Mr. K...    male  26.0      0      0          111369  30.0000  C148        C
890          891         0       3  Dooley, Mr....    male  32.0      0      0          370376   7.7500   NaN        Q

[891 rows x 12 columns]

shape: DataFrame의 행과 열 크기를 알아보는 방법, 행과 열을 튜플 형태로 반환


info(), describe(): 데이터의 다양한 정보를 알 수 있음

In [46]:
print('DataFrame 크기:', titanic_df.shape)

DataFrame 크기: (891, 12)


In [47]:
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


info: 데이터 건수, 데이터 타입, Null 검수

describe: 칼럼별 숫자형 데이터값의 n-percentile 분포도, 평균값, 최댓값, 최솟값 (*오직 숫자형)

**데이터의 분포도를 아는 것은 머신러닝 알고리즘의 성능을 향상시키는 중요한 요소**

In [48]:
titanic_df.describe()

PassengerId    Survived      Pclass         Age       SibSp       Parch        Fare
count   891.000000  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean    446.000000    0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std     257.353842    0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min       1.000000    0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%     223.500000    0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%     446.000000    0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%     668.500000    1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max     891.000000    1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

count는 Not Null인 데이터 건수

PassengerID 칼럼은 승객 ID를 식별 ( 1~891 )
: 분석에 의미 없는 속성
이런 경우에는 Series 객체에 value_counts( ) 사용: 해당 칼럼의 유형과 건수 확인

In [49]:
value_counts=titanic_df['Pclass'].value_counts()
print(value_counts)

3    491
1    216
2    184
Name: Pclass, dtype: int64


In [50]:
titanic_pclass=titanic_df['Pclass']
print(type(titanic_pclass))

<class 'pandas.core.series.Series'>


index값과 데이터 값을 출력
- 모든 Series와 DataFrame은 반드시 인덱스를 가짐

In [51]:
titanic_pclass.head(5)

0    3
1    1
2    3
3    1
4    3
Name: Pclass, dtype: int64

value_counts( ): Series 객체에서만 정의, DataFrame은 가지고 있지 않음

In [52]:
value_counts=titanic_df['Pclass'].value_counts()
print(type(value_counts))
print(value_counts)

<class 'pandas.core.series.Series'>
3    491
1    216
2    184
Name: Pclass, dtype: int64


*이때 인덱스가 0부터 시작 X

- 3, 1, 2: 인덱스는 단순히 순차 값과 같은 식별자만 할당하는 것이 아니라 고유성이 보장된다면 의미 있는 데이터값 할당 가능

- values_counts( )가 칼럼 값별 데이터 건수를 반환하므로 고유 칼럼 값을 식별자로 사용(단, 고유성 보장)

**DataFrame과 리스트, 딕셔너리, 넘파이 ndarray 상호 변환**

csv 파일을 DataFrame으로 생성, 이는 리스트, 딕셔너리, 넘파이 ndarray 등과 상호 변환 가능


**넘파이 ndarray, 리스트, 딕셔너리를 DataFrame으로 변환하기**

- DataFrame은 칼럼명을 가지고 있어 데이터 핸들링이 편함

- DataFrame으로 변환 시 칼럼명을 지정

- DataFrame은  행과 열을 가지는 2차원 데이터

In [53]:
import numpy as np

col_name1=['col1']
list1=[1,2,3]
array1=np.array(list1)
print('array1 shape:', array1.shape)

df_list1=pd.DataFrame(list1, columns=col_name1)
print('1차원 리스트로 만든 DataFrame:\n', df_list1)

df_array1=pd.DataFrame(array1, columns=col_name1)
print('1차원 ndarray로 만든 DataFrame:\n', df_array1)

array1 shape: (3,)
1차원 리스트로 만든 DataFrame:
    col1
0     1
1     2
2     3
1차원 ndarray로 만든 DataFrame:
    col1
0     1
1     2
2     3


In [54]:
col_name2=['col1', 'col2', 'col3']

list2=[[1,2,3],
       [11,12,13]]
array2=np.array(list2)
print('array2 shape:', array2.shape)
df_list2=pd.DataFrame(list2, columns=col_name2)
print('2차원 리스트로 만든 DataFrame:\n', df_list2)
df_array2=pd.DataFrame(array2, columns=col_name2)
print('2차원 ndarray로 만든 DataFrame:\n', df_array2)

array2 shape: (2, 3)
2차원 리스트로 만든 DataFrame:
    col1  col2  col3
0     1     2     3
1    11    12    13
2차원 ndarray로 만든 DataFrame:
    col1  col2  col3
0     1     2     3
1    11    12    13


**딕셔너리 -> DataFrame**

딕셔너리의 키 = 칼럼명


딕셔너리의 값 = 칼럼 데이터

In [55]:
dict={'col1':[1,11], 'col2':[2,22], 'col3':[3,33]}
df_dict=pd.DataFrame(dict)
print('딕셔너리로 만든 DataFrame:\n', df_dict)

딕셔너리로 만든 DataFrame:
    col1  col2  col3
0     1     2     3
1    11    22    33


DataFrame을 넘파이 ndarray, 리스트, 딕셔너리로 변환

In [56]:
#DataFrame을 ndarray로 변환
array3=df_dict.values
print('df_dict.values 타입:', type(array3), 'df_dict,values shape:', array3.shape)
print(array3)

df_dict.values 타입: <class 'numpy.ndarray'> df_dict,values shape: (2, 3)
[[ 1  2  3]
 [11 22 33]]


In [57]:
#DataFrame을 리스트로 변환
list3=df_dict.values.tolist()
print('df_dict,values.tolist() 타입:', type(list3))
print(list3)

#DataFrame을 딕셔너리로 변환
dict3=df_dict.to_dict('list')
print('\n df_dict.to_dict() 타입:', type(dict3))
print(dict3)

df_dict,values.tolist() 타입: <class 'list'>
[[1, 2, 3], [11, 22, 33]]

 df_dict.to_dict() 타입: <class 'dict'>
{'col1': [1, 11], 'col2': [2, 22], 'col3': [3, 33]}


**DataFrame의 칼럼 데이터 세트 생성과 수정**

In [58]:
#[ ] 연산자 이용

titanic_df['Age_0']=0
titanic_df.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked  Age_0
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S      0
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C      0
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S      0

새로운 칼럼명 Age_0으로 모든 데이터 값이 0으로 할당된 Series가 추가됨

In [59]:
titanic_df['Age_by_10']=titanic_df['Age']*10
titanic_df['Family_No']=titanic_df['SibSp']+titanic_df['Parch']+1
titanic_df.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked  Age_0  Age_by_10  Family_No
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S      0      220.0          2
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C      0      380.0          2
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S      0      260.0          1

기존 칼럼을 가공하여 새로운 칼럼 추가

In [60]:
titanic_df['Age_by_10']=titanic_df['Age_by_10']+100
titanic_df.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked  Age_0  Age_by_10  Family_No
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S      0      320.0          2
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C      0      480.0          2
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S      0      360.0          1

**DataFrame 데이터 삭제**

drop( ): DataFrame.drop(labels=None, axis=0, index=None, columns=None, level=None, inplace=False, errors='raise')

**주요 파라미터**
- labels
- axis: 특정 칼럼 또는 특정 행 드롭, 0=row, 1=column
- inplace: 디폴트가 False, Fase= 자기 자신의 DataFrame의 데이터는 삭제하지 않음, 수행 결과로 반환된 DataFrame에만 데이터 삭제

In [61]:
titanic_drop_df=titanic_df.drop('Age_0', axis=1)
titanic_drop_df.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked  Age_by_10  Family_No
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S      320.0          2
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C      480.0          2
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S      360.0          1

In [62]:
titanic_df.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked  Age_0  Age_by_10  Family_No
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S      0      320.0          2
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C      0      480.0          2
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S      0      360.0          1

In [63]:
drop_result=titanic_df.drop(['Age_0','Age_by_10','Family_No'], axis=1, inplace=True)
print('inplace=True로 drop 후 반환된 값:', drop_result)
titanic_df.head(3)

inplace=True로 drop 후 반환된 값: None


PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S

In [64]:
#로우 삭제

pd.set_option('display.width', 1000)
pd.set_option('display.max_colwidth',15)
print('#### before axis 0 drop ####')
print(titanic_df.head(3))

#### before axis 0 drop ####
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S


In [65]:
titanic_df.drop([0,1,2], axis=0, inplace=True)

print('#### after axis 0 drop ####')
print(titanic_df.head(3))

#### after axis 0 drop ####
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
3            4         1       1  Futrelle, M...  female  35.0      1      0  113803  53.1000  C123        S
4            5         0       3  Allen, Mr. ...    male  35.0      0      0  373450   8.0500   NaN        S
5            6         0       3  Moran, Mr. ...    male   NaN      0      0  330877   8.4583   NaN        Q


axis: 0=로우 삭제, 1=칼럼 삭제
inplace: 원본 유지 여부

**Index 객체**

index 객체는 레코드를 고유하게 식별하는 객체

In [66]:
#원본 파일 다시 로딩
titanic_df=pd.read_csv('titanic_train.csv')
#Index 객체 추출
indexes = titanic_df.index
print(indexes)
#Index 객체를 실제 값 array로 변환
print('Index 객체 array값:\n', indexes.values)

RangeIndex(start=0, stop=891, step=1)
Index 객체 array값:
 [  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17
  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35
  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53
  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71
  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89
  90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143
 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161
 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179
 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197
 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215
 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232

Index 객체는 1차원 array

단일 값 반환 및 슬라이싱 가능

In [67]:
print(type(indexes.values))
print(indexes.values.shape)
print(indexes[:5].values)
print(indexes.values[:5])
print(indexes[6])

<class 'numpy.ndarray'>
(891,)
[0 1 2 3 4]
[0 1 2 3 4]
6


*한 번 만들어진 DataFrame 및 Series의 Index 객체는 함부로 변경 불가능

In [68]:
indexes[0]=5
#변경 불가능

TypeError: ignored

In [69]:
series_fair=titanic_df['Fare']
print('Fair Series max 값:', series_fair.max())
print('Fair Series sum 값:', series_fair.sum())
print('sum() Fair Series:', sum(series_fair))
print('Fair Series +3 :\n', (series_fair +3).head(3))

Fair Series max 값: 512.3292
Fair Series sum 값: 28693.9493
sum() Fair Series: 28693.949299999967
Fair Series +3 :
 0    10.2500
1    74.2833
2    10.9250
Name: Fare, dtype: float64


In [70]:
titanic_reset_df=titanic_df.reset_index(inplace=False)
titanic_reset_df.head(3)

index  PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0      0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1      1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2      2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S

- 기존 index는 index라는 칼럼으로 추가
- reset_index( )는 인덱스가 연속된 int 숫자형 데이터가 아닐 경우 다시 이를 연속 int 숫자형 데이터로 만들 때 주로 사용

*Series의 reset_index( )를 적용하면 Series가 아닌 DataFrame이 반환

In [71]:
print( '### before reset_index ###')
value_counts=titanic_df['Pclass'].value_counts()
print(value_counts)
print('value_counts 객체 변수 타입:', type(value_counts))
new_value_counts=value_counts.reset_index(inplace=False)
print('### AFter reset_index ###')
print(new_value_counts)
print('new_value_couns 객체 변수 타입:', type(new_value_counts))

### before reset_index ###
3    491
1    216
2    184
Name: Pclass, dtype: int64
value_counts 객체 변수 타입: <class 'pandas.core.series.Series'>
### AFter reset_index ###
   index  Pclass
0      3     491
1      1     216
2      2     184
new_value_couns 객체 변수 타입: <class 'pandas.core.frame.DataFrame'>


**데이터 셀렉션 및 필터링**

- 판다스와 넘파이의 셀렉션과 필터링을 혼동하지 않게 주의
- ix[ ], iloc[ ], loc[ ] 

**DataFrame의 [ ] 연산자**


[ ]안에 들어갈 수 있는 것은 칼럼 명 문자 또는 인덱스로 변환 가능한 표현식(칼럼 지정 연산자)


In [73]:
print('단일 칼럼 데이터 추출:\n', titanic_df['Pclass'].head(3))
print('\n여러 칼럼의 데이터 추출:\n', titanic_df[['Survived', 'Pclass']].head(3))
print('[ ]안에 숫자 index는 KeyError 오류 발생:\n', titanic_df[0])

단일 칼럼 데이터 추출:
 0    3
1    1
2    3
Name: Pclass, dtype: int64

여러 칼럼의 데이터 추출:
    Survived  Pclass
0         0       3
1         1       1
2         1       3


KeyError: ignored

In [74]:
#슬라이싱 이용
titanic_df[0:2]

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch     Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0  A/5 21171   7.2500   NaN        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0   PC 17599  71.2833   C85        C

In [75]:
#불린 인덱싱
titanic_df[titanic_df['Pclass']==3].head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket   Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171  7.250   NaN        S
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...  7.925   NaN        S
4            5         0       3  Allen, Mr. ...    male  35.0      0      0          373450  8.050   NaN        S

- DataFrame 바로 뒤의 [ ] 연산자는 넘파이의 [ ]나 Series의 [ ]와 다름
- DataFrame 바로 뒤의 [ ] 내 입력 값은 칼럼명 또는 칼럼의 리스트를 지정해 칼럼 지정 연산에 사용하거나 불린 인덱스 용도로만
- DataFrame[0:2]와 같은 슬라이싱 연산으로 데이터 추출하는 방법은 사용하지 말자


**DataFrame ix[ ] 연산자**

ix[ ] 를 이용하면 원하는 위치의 데이터를 추출

**현재 ix연산자 X**

**칼럼 -> 문자: loc, 숫자: iloc**


In [78]:
print('칼럼 위치 기반 인덱싱 데이터 추출:', titanic_df.iloc[0,2])
print('칼럼 명 기반 인덱싱 데이터 추출:', titanic_df.loc[0,'Pclass'])

칼럼 위치 기반 인덱싱 데이터 추출: 3
칼럼 명 기반 인덱싱 데이터 추출: 3


In [79]:
data={'Name': ['Chulmin', 'Eunkyung', 'Jinwwong', 'Soobeom'],
      'Year': [2011, 2016, 2015, 2015],
      'Gender': ['Male',' Female', 'Male', 'Male']
      }
data_df=pd.DataFrame(data, index=['one','two','three','four'])
data_df                 

Name  Year   Gender
one     Chulmin  2011     Male
two    Eunkyung  2016   Female
three  Jinwwong  2015     Male
four    Soobeom  2015     Male

In [90]:
data_df.iloc[0,0]
#data_df.loc['one',0]
#data_df.loc[3,'Name']
data_df.iloc[0:2,[0,1]]
data_df.iloc[0:2, 0:3]
#data_df.loc[0:3, ['Name','Year']]
data_df.iloc[:]

Name  Year   Gender
one     Chulmin  2011     Male
two    Eunkyung  2016   Female
three  Jinwwong  2015     Male
four    Soobeom  2015     Male

**명칭 기반 인덱싱과 위치 기반 인덱싱의 구분**

iloc[ ] vs loc[ ]

In [96]:
data_df_reset=data_df.reset_index()
data_df_reset=data_df_reset.rename(columns={'index': 'old_index'})

data_df_reset.index=data_df_reset.index+1
data_df_reset

old_index      Name  Year   Gender
1       one   Chulmin  2011     Male
2       two  Eunkyung  2016   Female
3     three  Jinwwong  2015     Male
4      four   Soobeom  2015     Male

In [99]:
data_df_reset.ix[1,1]

#현재 ix를 사용할 수 없음

AttributeError: ignored

**DAtaFrame iloc[ ] 연산자**

In [100]:
data_df.iloc[0,0]

'Chulmin'

In [101]:
data_df.iloc[0, 'Name']

ValueError: ignored

In [102]:
data.df.iloc['one',0]

AttributeError: ignored

In [103]:
data_df_reset.iloc[0,1]

'Chulmin'

위처럼 문자는 위치 기반이 아니므로 값으로 받지 않고 불린 인덱싱도 제공하지 않음

**DataFrame loc[ ]연산자**

In [104]:
data_df.loc['one', 'Name']

'Chulmin'

In [106]:
data_df_reset.loc[1,'Name']

'Chulmin'

In [107]:
#오류나는 코드
data_df_reset.loc[0,'Name']

KeyError: ignored

In [110]:
#print('명칭 기반 ix slicing\n', data_df.ix['one':'two', 'NAme'], '\n')
#ix 지원X
print('위치 기반 iloc slicing\n', data_df.iloc[0:1,0],'\n')
print('명칭 기반 loc slicing\n', data_df.loc['one':'two','Name'])

위치 기반 iloc slicing
 one    Chulmin
Name: Name, dtype: object 

명칭 기반 loc slicing
 one     Chulmin
two    Eunkyung
Name: Name, dtype: object


*주의
loc[ ] 에서 정수를 사용하면 명칭 기반이므로 인덱스가 1, 2인 2개의 데이터를 반환

In [111]:
#print(data_df.ix[1:2, 'Name'])

**불린 인덱싱**

ix[ ]와 loc[ ]에서 불린 인덱싱 사용 가능

iloc[ ]에서는 불린 인덱싱 사용 불가

In [112]:
titanic_df=pd.read_csv('titanic_train.csv')
titanic_boolean=titanic_df[titanic_df['Age']>60]
print(type(titanic_boolean))
titanic_boolean

<class 'pandas.core.frame.DataFrame'>


PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch       Ticket      Fare        Cabin Embarked
33            34         0       2  Wheadon, Mr...    male  66.0      0      0   C.A. 24579   10.5000          NaN        S
54            55         0       1  Ostby, Mr. ...    male  65.0      0      1       113509   61.9792          B30        C
96            97         0       1  Goldschmidt...    male  71.0      0      0     PC 17754   34.6542           A5        C
116          117         0       3  Connors, Mr...    male  70.5      0      0       370369    7.7500          NaN        Q
170          171         0       1  Van der hoe...    male  61.0      0      0       111240   33.5000          B19        S
252          253         0       1  Stead, Mr. ...    male  62.0      0      0       113514   26.5500          C87        S
275          276         1       1  Andrews, Mi...  female  63.0      1      0        13502   77.9583           D7        S
280          281         0       3  Duane, Mr. ...    male  65.0      0      0       336439    7.7500          NaN        Q
326          327         0       3  Nysveen, Mr...    male  61.0      0      0       345364    6.2375          NaN        S
438          439         0       1  Fortune, Mr...    male  64.0      1      4        19950  263.0000  C23 C25 C27        S
456          457         0       1  Millet, Mr....    male  65.0      0      0        13509   26.5500          E38        S
483          484         1       3  Turkula, Mr...  female  63.0      0      0         4134    9.5875          NaN        S
493          494         0       1  Artagaveyti...    male  71.0      0      0     PC 17609   49.5042          NaN        C
545          546         0       1  Nicholson, ...    male  64.0      0      0          693   26.0000          NaN        S
555          556         0       1  Wright, Mr....    male  62.0      0      0       113807   26.5500          NaN        S
570          571         1       2  Harris, Mr....    male  62.0      0      0  S.W./PP 752   10.5000          NaN        S
625          626         0       1  Sutton, Mr....    male  61.0      0      0        36963   32.3208          D50        S
630          631         1       1  Barkworth, ...    male  80.0      0      0        27042   30.0000          A23        S
672          673         0       2  Mitchell, M...    male  70.0      0      0   C.A. 24580   10.5000          NaN        S
745          746         0       1  Crosby, Cap...    male  70.0      1      1    WE/P 5735   71.0000          B22        S
829          830         1       1  Stone, Mrs....  female  62.0      0      0       113572   80.0000          B28      NaN
851          852         0       3  Svensson, M...    male  74.0      0      0       347060    7.7750          NaN        S

In [113]:
#60세 이상인 승객의 나이와 이름만 추출하고 화면에 3개만 추출
titanic_df[titanic_df['Age']>60][['Name', 'Age']].head(3)

Name   Age
33  Wheadon, Mr...  66.0
54  Ostby, Mr. ...  65.0
96  Goldschmidt...  71.0

In [115]:
titanic_df.loc[titanic_df['Age']>60, ['Name', 'Age']].head(3)

Name   Age
33  Wheadon, Mr...  66.0
54  Ostby, Mr. ...  65.0
96  Goldschmidt...  71.0

In [116]:
#복합 연산자 사용
# and=& or=| not=~

titanic_df[(titanic_df['Age']>60) & (titanic_df['Pclass']==1) & (titanic_df['Sex']=='female')]

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
275          276         1       1  Andrews, Mi...  female  63.0      1      0   13502  77.9583    D7        S
829          830         1       1  Stone, Mrs....  female  62.0      0      0  113572  80.0000   B28      NaN

# **정렬, Aggregation 함수, GroupBy 적용**

**DataFrame, Series의 정렬 - sort_values( )**
- 주요 입력 파라미터: by, ascending, inplace
- by: 특정 칼럼 입력, 해당 칼럼으로 정렬을 수행
- ascending: True=오름차순 False=내림차순
- inplace: False= sortvalues( )를 호출한 DataFrame은 그대로 유지 True: 원본도 변경


In [117]:
titanic_sorted=titanic_df.sort_values(by=['Pclass','Name'], ascending=False)
titanic_sorted.head(3)

PassengerId  Survived  Pclass            Name   Sex   Age  SibSp  Parch    Ticket  Fare Cabin Embarked
868          869         0       3  van Melkebe...  male   NaN      0      0    345777   9.5   NaN        S
153          154         0       3  van Billiar...  male  40.5      0      2  A/5. 851  14.5   NaN        S
282          283         0       3  de Pelsmaek...  male  16.0      0      0    345778   9.5   NaN        S

**Aggregation 함수 적용**

DataFrame에서 바로 aggregation을 호출할 경우 모든 칼럼에 해당 aggregation을 적용

In [118]:
titanic_df.count()

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64

In [119]:
titanic_df[['Age','Fare']].mean()

Age     29.699118
Fare    32.204208
dtype: float64

**groupby() 적용**

파라미터 by에 칼럼을 입력하면 대상 칼럼으로 groupby

In [120]:
titanic_groupby=titanic_df.groupby(by='Pclass')
print(type(titanic_groupby))

<class 'pandas.core.groupby.generic.DataFrameGroupBy'>


In [ ]:
# groupby를 호출한 결과에 aggregation 함수를 호출하면 대상 칼럼 제외한 모든 칼럼에 aggergation 함수 적용

In [121]:
titanic_groupby=titanic_df.groupby('Pclass').count()
titanic_groupby

PassengerId  Survived  Name  Sex  Age  SibSp  Parch  Ticket  Fare  Cabin  Embarked
Pclass                                                                                    
1               216       216   216  216  186    216    216     216   216    176       214
2               184       184   184  184  173    184    184     184   184     16       184
3               491       491   491  491  355    491    491     491   491     12       491

In [122]:
# groupby 호출 후 count를 수행하면 groupby의 칼럼이 기준
titanic_groupby=titanic_df.groupby('Pclass')[['PassengerId','Survived']].count()
titanic_groupby

PassengerId  Survived
Pclass                       
1               216       216
2               184       184
3               491       491

In [124]:
#agg( )에 적용

titanic_df.groupby('Pclass')['Age'].agg([max,min])

max   min
Pclass            
1       80.0  0.92
2       70.0  0.67
3       74.0  0.42

In [125]:
# 딕셔너리 형태로 유연하게
# 적용될 칼럼과 함수
agg_format={'Age':'max', 'SibSp':'sum','Fare':'mean'}
titanic_df.groupby('Pclass').agg(agg_format)

Age  SibSp       Fare
Pclass                        
1       80.0     90  84.154687
2       70.0     74  20.662183
3       74.0    302  13.675550

**결손 데이터 처리하기**

판다스는 결손 데이터를 처리하는 API 제공

결손 데이터=NULL=NaN

NaN 값은 평균, 총합 등 연산 시 제외

- isna(): NaN 여부 확인
- fillna(): NaN 값을 다른 값으로 대체

In [126]:
titanic_df.isna().head(3)

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket   Fare  Cabin  Embarked
0        False     False   False  False  False  False  False  False   False  False   True     False
1        False     False   False  False  False  False  False  False   False  False  False     False
2        False     False   False  False  False  False  False  False   False  False   True     False

In [127]:
# sum을 이용하여 결손 데이터 개수 확인
titanic_df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

**fillna( )로 결손 데이터 대체하기**

In [128]:
titanic_df['Cabin']=titanic_df['Cabin'].fillna('C000')
titanic_df.head(3)

#inplace=True 를 해주어야 원본 데이터 적용

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500  C000        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250  C000        S

In [129]:
titanic_df['Age']=titanic_df['Age'].fillna(titanic_df['Age'].mean())
titanic_df['Embarked']=titanic_df['Embarked'].fillna('S')
titanic_df.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

**apply lambda 식으로 데이터 가공**

판다스가 속도 면에서 더욱 빠르지만 복잡한 데이터 가공이 필요한 경우 apply lambda를 이용

In [130]:
def get_square(a):
  return a**2

print('3의 제곱은:', get_square(3))

3의 제곱은: 9


lambda는 함수의 선언과 함수 내의 처리를 한 중릐 식으로 쉽게 변환

In [131]:
#위의 함수를 변환
lambda_square=lambda x: x**2
print('3의 제곱은:', lambda_square(3))

3의 제곱은: 9


In [133]:
# lambda 식에서 여러 개의 값을 입력 인자로 받을 경우 map( ) 함수를 결합

a=[1, 2, 3]
squares=map(lambda x: x**2 , a)
list(squares)

[1, 4, 9]

In [134]:
titanic_df['Name_len']=titanic_df['Name'].apply(lambda x: len(x))
titanic_df[['Name', 'Name_len']].head(3)

Name  Name_len
0  Braund, Mr....        23
1  Cumings, Mr...        51
2  Heikkinen, ...        22

In [136]:
# 나이 15세를 기준으로 Child, Adult 구분

titanic_df['Child_Adult']=titanic_df['Age'].apply(lambda x: 'Child' if x<=15 else 'Adult')
titanic_df[['Age', 'Child_Adult']].head(8) 

Age Child_Adult
0  22.000000       Adult
1  38.000000       Adult
2  26.000000       Adult
3  35.000000       Adult
4  35.000000       Adult
5  29.699118       Adult
6  54.000000       Adult
7   2.000000       Child

* if esle를 lambda에서 사용할 때에는 위치가 중요!

if는 반환값 뒤에 else는 앞에

In [138]:
# else if 적용하기
titanic_df['Age_cat']=titanic_df['Age'].apply(lambda x: 'Child' if x<=15 else ('Adult' if x <=60 else 'Elderly'))
titanic_df['Age_cat'].value_counts()

Adult      786
Child       83
Elderly     22
Name: Age_cat, dtype: int64

In [141]:
#함수를 사용하여 더욱 세분화된 분류

def get_category(age):
  cat=''
  if age<=5: cat='baby'
  elif age<=12: cat='child'
  elif age<=18: cat='Teenager'
  elif age<=25: cat='Student'
  elif age<=35: cat='Young Adult'
  elif age<=60: cat='Adult'
  else: cat='Elderly'

  return cat

titanic_df['Age_cat']=titanic_df['Age'].apply(lambda x: get_category(x))
titanic_df[['Age', 'Age_cat']].head()

Age      Age_cat
0  22.0      Student
1  38.0        Adult
2  26.0  Young Adult
3  35.0  Young Adult
4  35.0  Young Adult

### **05 정리**

---

파이썬의 대표적인 시각화 패키지인 맷플롯립과 시본의 사용법을 머신러닝 프로그램을 직접 만들어 보면서 학습하는 것이 좋다.

모델을 생성하고 예측을 수행하는 데 있어서 데이터를 전처리하고 적절한 피처를 가공/추출하는 부분이 훨씬 많은 비중을 차지하므로 넘파이, 판다스, 맷플롯립/시본과 같은 파이썬 머신러닝 생태계를 이루고 있는 다양한 패키지에 대한 이해가 중요